From 90aa4cec5246b56daa9354ebf45723a3e396be57 Mon Sep 17 00:00:00 2001
From: Alexandru Croitor <alexandru.croitor@qt.io>
Date: Wed, 29 Oct 2025 14:50:17 +0100
Subject: [PATCH 550/553] CMake: Fix order of Qt6*AdditionalTargetInfo.cmake
 file installation

Qt6*AdditionalTargetInfo.cmake files are defer installed using a
finalizer at the end of the directory scope where
qt_internal_export_additional_targets_file is called.
For tools packages, there are usually installed at the end of the root
directory scope.

qt_internal_export_additional_targets_file keeps track of which files
to install (based on an export id) in a global property.

Some tests might create a Mock Qt module which also calls the above
function and adds its own export id to the global property.

Because the mock module finalizer is called before the Tools one,
the mock finalizer will process the ids of both of the module and the
Tools package, ending up creating installation rules for the Tools
in the test directory scope.

When Qt is configured with
  -DQT_BUILD_TESTS=ON -DQT_BUILD_TESTS_BY_DEFAULT=OFF
test installation rules are skipped, which means the Tools files would
never be installed, causing configuration errors when looking up that
package later.

To avoid this situation, instead of collecting the export ids in a
global property to process all of them later, just let each finalizer
process a single id in the directory scope where the finalizer is
called.

Uncovered by d66cfa93eefa185aa15d42e9dc38cbcdbd0578cb in
qtdeclarative.

Pick-to: 6.8
Fixes: QTBUG-141466
Change-Id: Icf839835ea028df2e0079995e301693b9c3cd342
Reviewed-by: Joerg Bornemann <joerg.bornemann@qt.io>
(cherry picked from commit 682c8ecd3227e53812242ed50f74feec0b7f88eb)
Reviewed-by: Qt Cherry-pick Bot <cherrypick_bot@qt-project.org>
---
 cmake/QtTargetHelpers.cmake | 19 ++++---------------
 1 file changed, 4 insertions(+), 15 deletions(-)

diff --git a/cmake/QtTargetHelpers.cmake b/cmake/QtTargetHelpers.cmake
index 632fb5b5644..7b2d51d69a7 100644
--- a/cmake/QtTargetHelpers.cmake
+++ b/cmake/QtTargetHelpers.cmake
@@ -841,13 +841,12 @@ function(qt_internal_export_additional_targets_file)
 
     qt_internal_append_export_additional_targets()
 
-    set_property(GLOBAL APPEND PROPERTY _qt_export_additional_targets_ids "${id}")
     set_property(GLOBAL APPEND
         PROPERTY _qt_export_additional_targets_export_name_prefix_${id} "${arg_EXPORT_NAME_PREFIX}")
     set_property(GLOBAL APPEND
         PROPERTY _qt_export_additional_targets_config_install_dir_${id} "${arg_CONFIG_INSTALL_DIR}")
 
-    qt_add_list_file_finalizer(qt_internal_export_additional_targets_file_finalizer)
+    qt_add_list_file_finalizer(qt_internal_export_additional_targets_file_finalizer "${id}")
 endfunction()
 
 function(qt_internal_get_export_additional_targets_id export_name out_var)
@@ -913,19 +912,9 @@ function(qt_internal_validate_export_additional_targets)
     set(arg_TARGET_EXPORT_NAMES "${arg_TARGET_EXPORT_NAMES}" PARENT_SCOPE)
 endfunction()
 
-# The finalizer might be called multiple times in the same scope, but only the first one will
-# process all the ids.
-function(qt_internal_export_additional_targets_file_finalizer)
-    get_property(ids GLOBAL PROPERTY _qt_export_additional_targets_ids)
-
-    foreach(id ${ids})
-        qt_internal_export_additional_targets_file_handler("${id}")
-    endforeach()
-
-    set_property(GLOBAL PROPERTY _qt_export_additional_targets_ids "")
-endfunction()
-
-function(qt_internal_export_additional_targets_file_handler id)
+# The finalizer might be called multiple times in the same directory scope, but it will only process
+# one specific id.
+function(qt_internal_export_additional_targets_file_finalizer id)
     get_property(arg_EXPORT_NAME_PREFIX GLOBAL PROPERTY
         _qt_export_additional_targets_export_name_prefix_${id})
     get_property(arg_CONFIG_INSTALL_DIR GLOBAL PROPERTY
-- 
2.51.2

